src/tap/src/tapcell.cpp

```diff
@@ -33,41 +33,1306 @@
+
+  odb::dbBox* inst_bb = inst->getBBox();
+  vector<int> filled_sites_vector;
+  filled_sites_vector.push_back(inst_bb->yMin());
+  filled_sites_vector.push_back(inst_bb->xMin());
+  filled_sites_vector.push_back(inst_bb->xMax());
+  filled_sites_.push_back(filled_sites_vector);
+
+  phy_idx_++;
+}
+
+void Tapcell::buildRow(odb::dbBlock* block,
+                       string name,
+                       odb::dbSite* site,
+                       int start_x,
+                       int end_x,
+                       int y,
+                       odb::dbOrientType& orient,
+                       odb::dbRowDir& direction,
+                       int& min_row_width)
+{
+  int site_width = site->getWidth();
+
+  int new_row_num_sites = (end_x - start_x) / site_width;
+  int new_row_width = new_row_num_sites * site_width;
+
+  if (new_row_num_sites > 0 && new_row_width >= min_row_width) {
+    odb::dbRow::create(block,
+                       name.c_str(),
+                       site,
+                       start_x,
+                       y,
+                       orient,
+                       direction,
+                       new_row_num_sites,
+                       site_width);
+  }
+}
+
+vector<vector<odb::dbRow*>> Tapcell::organizeRows()
+{
+  std::map<int, vector<odb::dbRow*>> rows_dict;
+  for (odb::dbRow* row : db_->getChip()->getBlock()->getRows()) {
+    odb::Rect rowBB;
+    row->getBBox(rowBB);
+    if (rows_dict.find(rowBB.yMin()) == rows_dict.end()) {
+      vector<odb::dbRow*> rows_vector;
+      rows_vector.push_back(row);
+      rows_dict.insert(
+          std::pair<int, vector<odb::dbRow*>>(rowBB.yMin(), rows_vector));
+    } else {
+      rows_dict[rowBB.yMin()].push_back(row);
+    }
+  }
+
+  // organize rows bottom to top
+  vector<vector<odb::dbRow*>> rows;
+  for (std::map<int, vector<odb::dbRow*>>::iterator iter = rows_dict.begin();
+       iter != rows_dict.end();
+       ++iter) {
+    int k = iter->first;
+    std::map<int, odb::dbRow*> in_row_dict;
+    for (odb::dbRow* in_row : rows_dict[k]) {
+      odb::Rect row_BB;
+      in_row->getBBox(row_BB);
+      in_row_dict.insert(std::pair<int, odb::dbRow*>(row_BB.xMin(), in_row));
+    }
+    // organize sub rows left to right
+    vector<odb::dbRow*> in_row;
+    for (std::map<int, odb::dbRow*>::iterator iter = in_row_dict.begin();
+         iter != in_row_dict.end();
+         ++iter) {
+      int in_k = iter->first;
+      in_row.push_back(in_row_dict[in_k]);
+    }
+    rows.push_back(in_row);
+    in_row.clear();
+  }
+
+  return rows;
+}
+
+int Tapcell::removeCells(const char* prefix)
+{
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  string prefix_string(prefix);
+  int removed = 0;
+
+  if (prefix_string.length() == 0) {
+    // if no prefix is given, return 0 instead of having all cells removed
+    return 0;
+  }
+
+  for (odb::dbInst* inst : block->getInsts()) {
+    if (inst->getName().find(prefix_string) != string::npos) {
```

Comment:
Are you really intending to find the prefix anywhere in the string?  If so it isn't really a prefix.  If not, use substr to get the string prefix for comparison.

Reply:
What I understand it is a prefix then and the condition just checks if it's included in the name or not, should something be changed?

Reply:
"abc".find("b") will != npos because "b" is somewhere in the string, not just at the start (ie a prefix).

Reply:
You can use == 0 to test that it is at the start rather than != string::npos